File: butcher/a_greedy_beggar.pl
my $FollowTarget;
my $User;

sub EVENT_AGGRO {
 	quest::say("Guards! Guards! Help me!!")
}

sub EVENT_SAY {
	#:: Match if FollowTarget is not defined (nobody has triggered a follow event yet)
	if (!$FollowTarget) {
		quest::say("Hello there $name. I don't suppose you can spare some coins? I'm just a poor halfling that is far away from home. I can't afford anything to eat or drink. Anything you can offer me will be of help.");
		$FollowTarget = $charid;
		$User = $userid;
		#:: Set a timer to loop every 5 seconds
		quest::settimer("follow", 5);
		#:: Set appearance to Stand
		$npc->SetAppearance(0);
	}
}

sub EVENT_ITEM {
	#:: Match if FollowTarget is not defined (nobody has triggered a follow event yet)
	if (!$FollowTarget) {
		quest::say("Oh thank you . You are too kind to this poor halfling. Do you have anything else to give me?");
		$FollowTarget = $charid;
		$User = $userid;
		#:: Set a timer to loop every 5 seconds
		quest::settimer("follow", 5);
		#:: Set appearance to Stand
		$npc->SetAppearance(0);
	}
}

sub EVENT_TIMER {
	#:: Match timer "follow"
	if ($timer eq "follow") {
		#:: Match if FollowTarget is defined (someone triggered a follow event)
		if ($FollowTarget) {
			#:: Create a scalar variable to store the following target by entity ID
			my $FollowingTarget = $entity_list->GetClientByCharID($FollowTarget);
			#:: Match if the target was defined, but is no longer on the entity list
			if (!$FollowingTarget) {
				undef $FollowTarget;
				undef $User;
				#:: Stop following (just in case)
				quest::sfollow();
				#:: Return to spawn point
				quest::moveto(2407, 1482, -0.25, 331, 1);
			}
			#:: Match if the following target is on the entity list
			else {
				#:: Follow the target who triggered the follow event
				quest::follow($User);
			}
		}
		#:: Match if no follow target is defined
		else {
			#:: Match if back at spawn point
			if ($x == 2407 && $y == 1482) {
				#:: Set appearance to Feign Death
				$npc->SetAppearance(3);
				#:: Cleanup the timer "follow"
				quest::stoptimer("follow");
			}
		}
	}
}

File: butcher/a_greedy_beggar.lua
local followtarget;
local zonecheck;

function event_spawn(e)
	e.self:Say("event spawn triggered.");
	if(followtarget == nil) then
		e.self:Say("Followtarget is nil.");
	else
		e.self:Say(string.format("Followtarget is %s", followtarget));
	end
end

function event_say(e)
	e.self:Say("event say triggered.");
	if(e.message:findi("hail")) then
		e.self:Say("Hello there " .. e.other:GetName() .. ". I don't suppose you can spare some coins? I'm just a poor halfling that is far away from home. I can't afford anything to eat or drink. Anything you can offer me will be of help.");  -- response to all messages, even when following
		if(followtarget == nil) then
			followtarget = e.other:GetID();
			zonecheck = 1;
			e.self:Say(string.format("Followtarget is %s", followtarget));
			e.self:SetAppearance(0); -- Stand
			eq.set_timer("follow",5000);
			e.self:Say("Timer 'follow' set.");
		end
	end
end

function event_timer(e)
	e.self:Say("event timer triggered.");
	if(zonecheck == 1) then
		if(eq.get_entity_list():GetClientByID(followtarget)) then
			e.self:Say("Follow target was assigned and is in zone.");
			e.self:Say(string.format("Followtarget is %s", followtarget));
			eq.follow(followtarget); -- Follow the player who triggered the event
			e.self:Say("following target.");
		else
			zonecheck = nil;
			e.self:Say(string.format("Followtarget is %s and he is not in zone", followtarget));
			eq.stop_follow();
			e.self:Say("Stopping follow.");
			followtarget = nil;
			e.self:Say("Followtarget is nil.");
			eq.stop_timer("follow");
			e.self:Say("Timer 'follow' stopped.");
			e.self:SetAppearance(3); -- lying down
		end
	else
		e.self:Say(string.format("Followtarget is %s and he is not in zone", followtarget));
		eq.stop_follow();
		e.self:Say("Stopping follow.");
		followtarget = nil;
		e.self:Say("Followtarget is nil.");
		eq.stop_timer("follow");
		e.self:Say("Timer 'follow' stopped.");
		e.self:SetAppearance(3); -- lying down
	end
end

function event_trade(e) -- Note that we are intentionally accepting any trade, as long as we are not following
	local item_lib = require("items");
	e.self:Say("event item triggered.");
	if(followtarget == nil or e.other:GetID() == followtarget) then
		e.self:Say("Followtarget is nil or the guy who just traded with me.");
		e.self:Say("Oh thank you. You are too kind to this poor halfling. Do you have anything else to give me?");
		e.self:SetAppearance(0); -- Stand
		followtarget = e.other:GetID();
		zonecheck = 1;
		e.self:Say(string.format("Followtarget is %s", followtarget));
		eq.set_timer("follow",5000);
		e.self:Say("Timer 'follow' set.");
	else
		item_lib.return_items(e.self, e.other, e.trade)
		e.self:Say(string.format("Followtarget is %s and I do not want your items", followtarget));
	end
end
